### ISSUE TITLE
[Bug]: Train Conformer ASR for Librispeech get bad results without CTC loss

### ISSUE THREAD

**Issue by can-cui** [speechbrain/speechbrain#1875]
### Describe the bug

Hi,
In the Librispeech recipe, I have tried training the Conformer ASR without CTC loss. So I set the ctc weight to 0 in the parameter file:
```
ctc_weight: 0.0
grad_accumulation_factor: 2
max_grad_norm: 5.0
loss_reduction: "batchmean"
sorting: descending
```
For the decoding parameters I have the following settings:
```
min_decode_ratio: 0.0
max_decode_ratio: 1.0
valid_search_interval: 10
valid_beam_size: 10
test_beam_size: 10
lm_weight: 0.60
ctc_weight_decode: 0
```
And I use ```!new:speechbrain.decoders.S2STransformerBeamSearch``` for validation and test search.
I trained 30 epochs on the Librispeech clean 100 data.
By observing the training curve, I found that the train loss was decreasing all the time during training, but the validation loss was not decreasing.
![output](https://user-images.githubusercontent.com/56673370/223988337-734a6a55-4ba1-49d8-8cf8-027a0830456b.png)
The final test result is also very bad : ```Epoch loaded: 30 - test loss: 1.09e+02, test ACC: 1.52e-01, test WER: 99.13```
I would like to ask if there are some settings in my decoding parameter that need to be changed to match the ctc weight of 0? 

### Expected behaviour

I would like to know why there is such a big difference in the loss of train and validation, hoping to achieve more normal training results for Conformer ASR without using ctc loss.

### To Reproduce

```
id=0,1,2,3
CUDA_VISIBLE_DEVICES=$id python train.py hparams/conformer_small.yaml --data_parallel_backend
```

**Comment by TParcollet** [comment]
Hi can you try with regularMHA for positional encoding and report the results here please?

**Comment by can-cui** [comment]
> Hi can you try with regularMHA for positional encoding and report the results here please?

Hi, I got similar results with this parameter:

![output1](https://user-images.githubusercontent.com/56673370/224245390-5d65b964-456b-4725-9d6c-c382349f9786.png)

```Epoch loaded: 30 - test loss: 2.21e+02, test ACC: 1.52e-01, test WER: 96.54```

**Comment by TParcollet** [comment]
The problem comes from the batch size, 12 is not big enough for attention to learn the alignement in that setup. You should try having an effective batch size around 100 either by increasing the grad accum or by increasing the BS. I would encourage you to use DDP instead of DP as well as encouraged by PyTorch.

**Comment by can-cui** [comment]
> The problem comes from the batch size, 12 is not big enough for attention to learn the alignement in that setup. You should try having an effective batch size around 100 either by increasing the grad accum or by increasing the BS. I would encourage you to use DDP instead of DP as well as encouraged by PyTorch.

Not easy to get to that value, but thanks for the reply!

**Comment by TParcollet** [comment]
grad_accum won't affect your VRAM consumption. So if you have a smallish GPU, just play with that. If it's with dynamic batching, indeed, things become a bit messy ... just here, just try to make sure that "most" batches are big enough: above 64.

**Comment by Adel-Moumen** [comment]
Hi, 

any news regarding this issue please? :)

**Comment by can-cui** [comment]
> Hi,
> 
> any news regarding this issue please? :)

Hi,

I reused a larger train set (Librispeech clean 100 + 360) and a larger grad_accum value (with batch size 4, 4 GPUs, gradient accumulation value 10) for no ctc training. I did get a better result and I found that in this case it was necessary to increase the number of heads of attention appropriately, I increased the number of heads from 4 to 8. Here are the experimental results：
![100360](https://user-images.githubusercontent.com/56673370/229075789-97e00126-b751-472c-89a3-ca1eefead4e4.png)

the test result is 14.65% for test clean set. Of course there is still a gap between this result and the expected result, and I am trying to find a solution from adjusting the training scheme (for example, changing the LR according to different epoch stages, because I found that the default LR schedule's don't change the LR very much)

**Comment by TParcollet** [comment]
Hi, do not forget to pull the latest develop. I fixed a bug with conformer.

**Comment by can-cui** [comment]
> Hi, do not forget to pull the latest develop. I fixed a bug with conformer.

Thanks for your reminder! I reran the same experiment on the latest version and after 100 epochs, I obtained a WER of 5.38%. The test results using CTC loss were 4.95%. I think I have achieved satisfactory results and I am very grateful for your patient help!

**Comment by AsuMagic** [comment]
I am assuming this issue is resolved so I will close it as such, feel free to reopen if I'm wrong.
